src/hww.c

```diff
@@ -149,28 +168,62 @@ static size_t _api_info(uint8_t* buf)
     return current - buf;
 }
 
+static void _clear_state(void)
+{
+    _state.blocking_op = HWW_BLOCKING_OP_NONE;
+    _state.unlock_wf = NULL;
+}
+
+static void _lock(hww_blocking_op_t op)
+{
+    _state.blocking_op = op;
+    usb_processing_lock(usb_processing_hww());
+}
+
+/**
+ * Clears the current state and unlocks the USB stack (if locked).
+ */
+static void _unlock(void)
+{
+    _clear_state();
+    if (_state.blocking_op != HWW_BLOCKING_OP_NONE) {
+        usb_processing_unlock();
+    }
+}
+
+static void _unlock_finished_cb(bool result, void* param)
```

Comment:
is this ever called with `result=false` on cancel? not sure i understand how HWW_REQ_CANCEL leads to the proper (negative) response.

Reply:
The `HWW_REQ_CANCEL` is handled by the separate arbitration logic, which workflows know nothing about. The only case the unlock can fail (`result == false`) is if `memory_is_initialized() == false` when the unlock workflow is started.

`HWW_REQ_CANCEL` is handled internally by `hww.c` (see `_process_msg`).

Reply:
you mean `_cancel_packet`. thx, i see now